### PR TITLE
Introduce typing for Variables and Collections

### PR DESCRIPTION
**Pull request by lindner**
- Allows for verifying Collections/Variables interface at compile time.
- Code can cast the result of construct/connect calls to either type.
- Cleans up existing usages/hacks.

### CODE REVIEW

**Comment by lindner** [comment]
Very much open to suggestions on how to make this better!

Or if you have something else in the works let me know and I'll drop this.

File: src/runtime/storage/pouchdb/pouch-db-storage.ts [hunk 1]
```diff
@@ -22,6 +22,7 @@ import PouchDbDebug from 'pouchdb-debug';
 
 PouchDB.plugin(PouchDbDebug);
 PouchDB.debug.disable();
+      PouchDB.debug.enable('*');
```
Comment (lindner) [inline src/runtime/storage/pouchdb/pouch-db-storage.ts]: extraneous...  (will revert)

File: src/runtime/storage/pouchdb/pouch-db-collection.ts [hunk 2]
```diff
@@ -1,9 +1,9 @@
 import {CrdtCollectionModel} from '../crdt-collection-model.js';
 import {assert} from '../../../platform/assert-web.js';
 import {PouchDbStorageProvider} from './pouch-db-storage-provider.js';
-import {Type} from '../../type.js';
```
Comment (lindner) [inline src/runtime/storage/pouchdb/pouch-db-collection.ts]: this belongs in another PR..

File: src/runtime/storage/storage-provider-base.ts [hunk 3]
```diff
@@ -19,6 +19,30 @@ enum EventKind {
...
+export interface CollectionStorageProvider extends StorageProviderBase {
+  // tslint:disable-next-line: no-any
+  toList(): Promise<any[]>;
+
+  // tslint:disable-next-line: no-any
+  getMultiple(ids: string[]): Promise<any[]>;
+  storeMultiple(values, keys: string[], originatorId: string): Promise<void>;
+
+  // tslint:disable-next-line: no-any
+  get(id: string): any;
```
Comment (shans) [inline src/runtime/storage/storage-provider-base.ts]: I think you can use {} instead of any, but then maybe this makes the call sites uglier.

File: src/runtime/storage/storage-provider-base.ts [hunk 4]
```diff
@@ -19,6 +19,30 @@ enum EventKind {
...
+export interface VariableStorageProvider extends StorageProviderBase {
+  // tslint:disable-next-line: no-any
+  get(): Promise<any>;
+  set(val: {}): Promise<void>;
+}
+
+/**
+ * Methods that must be implemented by a Collection Storage Provider
+ */
+export interface CollectionStorageProvider extends StorageProviderBase {
```
Comment (shans) [inline src/runtime/storage/storage-provider-base.ts]: also remove, removeMultiple

File: src/runtime/storage/storage-provider-base.ts [hunk 5]
```diff
@@ -19,6 +19,30 @@ enum EventKind {
 }
 type Callback = ({}) => void;
 
+/**
+ * Methods that must be implemented by a Variable Storage Provider
+ */
+export interface VariableStorageProvider extends StorageProviderBase {
```
Comment (shans) [inline src/runtime/storage/storage-provider-base.ts]: also clear

File: src/runtime/storage/storage-provider-base.ts (hunk 3 above)
Comment (lindner) [reply src/runtime/storage/storage-provider-base.ts]: seems to work fine!  Thanks for the suggestion.

Comment (lindner) [reply src/runtime/storage/storage-provider-base.ts]: I take that back :(   will find a better solution in a followup PR.

File: src/runtime/storage/storage-provider-base.ts (hunk 4 above)
Comment (lindner) [reply src/runtime/storage/storage-provider-base.ts]: done

File: src/runtime/storage/storage-provider-base.ts (hunk 5 above)
Comment (lindner) [reply src/runtime/storage/storage-provider-base.ts]: done
